### ISSUE TITLE
Override styles via React props and classnames library

### ISSUE THREAD

**Issue by flogy** [tailwindcss/discuss#360]
I am using **TailwindCSS** in a **ReactJS** project. As encouraged by the docs I'm following the utility-first approach and almost never extract a component. When I have repeating code I just create a React component out of it to remove duplications.

For these React components I like to provide an optional `className` prop, so additional TailwindCSS styles can be added from outside if required. I use the [classnames library](https://github.com/JedWatson/classnames) to join the passed in prop together with the pre-existing ones.

Here is an example:

_Link.jsx_
```
import * as React from "react";
import classNames from "classnames";

export const Link = props => <a className={classNames("m-2 p-2 bg-gray-200", props.className)}>{props.children}</a>;
```

_App.jsx_
```
import * as React from "react";
import {Link} from "./Link";

export const App = <Link className="bg-white">Hello world</Link>;
```

Despite expecting the background color of the `Link` to be white now, it is still gray. After some research I found out that this is [because "the order in which the attributes are overwritten is not determined by the order the classes are defined in the class attribute, but instead where they appear in the CSS."](https://stackoverflow.com/questions/1321692/how-to-specify-the-order-of-css-classes).

I was able to proof that it is not some other fault by passing in `bg-red` instead of `bg-white` which then actually changed the background color to red, as expected. When inspecting the HTML element and its CSS in Chrome, I could also see that `bg-red` seems to be overriding `bg-gray-200` which overrides `bg-white`.

**So now my questions are:**

1. Are there any workarounds / solutions to successfully override certain classnames if desired?
2. How do you guys handle such situations?

Thanks in advance!

**Comment by hacknug** [comment]
The project is not maintained right now but it might be helpful: https://github.com/bradlc/babel-plugin-tailwind-components

**Comment by flogy** [comment]
> The project is not maintained right now but it might be helpful: https://github.com/bradlc/babel-plugin-tailwind-components

Thanks for mentioning this. Although I already tried to combine TailwindCSS and CSS-in-JS libraries using this approach and failed to get it working, I have now tried my luck for a second time... and this time I made it, thanks to an [awesome Gatsby theme](https://github.com/talensjr/gatsby-theme-tailwindcss) that sets everything up!

After migrating the code it really worked as expected! 🎉

In case others get stuck with the same problem, I have created a CodeSandbox to demonstrate how it can be done: [Demo](https://codesandbox.io/s/gatsby-tailwind-emotion-demo-j8n0l?fontsize=14)
